### PR TITLE
Batching support and small fixes

### PR DESCRIPTION
**Pull request by neworderofjamie**
I've been having a go at using PyGeNN to train e-prop (http://www.nature.com/articles/s41467-020-17236-y) networks on SHD using Tonic and have fixed/implemented a few things:
- The data loader wasn't exported before
- I think ``DataLoader.__iter__`` should return a copy as otherwise you can't restart iterations at the beginning of a new epoch (perhaps it should also reshuffle the copy's indices if shuffle is required)
- The data loader didn't support batching. I've added preliminary support to the data loader and (for now) to the HSD base class 

If you're happy with the general direction (not knowing PyTorch, I'm not super-familiar with the original data-loader!), let me know and I'll implement batching in the other datasets and add tests.

### CODE REVIEW

**Comment by biphasic** [comment]
hello Jamie, thank you for submitting a PR! There's multiple things:
1. Supporting a custom DataLoader leads down a path that will be very expensive to maintain. This is my mistake, as I should not have added the custom DataLoader class in the first place. I would like to reuse a maximum of functionality that is already out there and just provide an interface for it. When it comes to PyTorch, we already have a nice batching support using their DataLoader (using a custom collate function).
2. The issue so far was the lacking support for TensorFlow (TF). I had another look and found a nice example [here](https://huggingface.co/docs/datasets/torch_tensorflow.html). I think the easiest option would be to use from_tensor_slices like [here](https://www.tensorflow.org/tutorials/load_data/numpy) to wrap the data in a TF class! The transforms would still have to be applied beforehand, but at least afterwards it will be compatible with what TF users are already familiar with. Let me know if you'd be happy with that solution or not!

**Comment by neworderofjamie** [comment]
Ahh ok, that makes sense. I'll start making a GeNN specific dataloader instead. Does the PyTorch data loader not expect dataset's ``__getitem__`` magic method to be able to handle lists of indices? Like I said, I'm trying to de-couple our GeNN-based ML library from Tensor Flow so I don't really have an opinion on how you handle TF 😄

**Comment by biphasic** [comment]
I see, there was a misunderstanding there, sorry about that. I would say that PT's DataLoader would be able to handle multiple indices or at least not break. However using a try/except block in the __getitem__ method is brittle and expensive, so I would rather avoid that. I see that [this](https://stackoverflow.com/questions/27954297/how-to-write-getitem-cleanly) Stack Overflow question is similar.

As I understand now you would like to have the ability to load multiple event samples at once so you can use your own data loader class afterwards that is independent of both PT and TF. But the issue of collating those samples that likely have different length is not a trivial one. The assumption to append them and return a list is a strong one, and does not add any performance benefits. The same can be done outside the __getitem__ method (by looping over indices) and saves the standard case the isinstance / try&except checks. Therefore I will not merge this for the moment, sorry

**Comment by neworderofjamie** [comment]
No worries, thanks for the feedback. Like I say, my knowledge of the PyTorch data loader is pretty limited so I didn't really know where responsibility should lie. Will close this PR.
